src/elexmodel/models/BootstrapElectionModel.py

```diff
@@ -0,0 +1,1340 @@
+        # the first coefficient is the median of all deltas where x = [1, 0] and the second coefficient
+        # is the median of all deltas where x = [0, 1]
+        # since the covariates define the strata, we get the tau-th (e.g. median or 30th percentile)
+        # delta per strata as the beta for that regression, which defines a probability distribution.
+
+        # for each stratum we want to add a worst case lower bound (for the taus between 0-0.49)
+        # and upper bound (for the taus between 0.5-1) so that if we sample a uniform from a different
+        # stratum that is larger/smaller than any one we have seen in this strata, we have a worst case
+        # value that is larger/smaller than what we saw in that strata. We do this by simply adding the
+        # lower/upper bound to the regression, one pair for each stratum.
+        # This lower/upper bound is set manually, note that if we observe a value that is more extreme than
+        # the lower/upper bound we are fine, since the quantile regression will use that instead
+        for x_stratum in x_strata:
+            x_train_aug = np.concatenate([x_train_strata, x_stratum.reshape(1, -1)], axis=0)
+
+            delta_aug_lb = np.concatenate([delta_hat, [lb]])
+            delta_aug_ub = np.concatenate([delta_hat, [ub]])
+
+            # fit the regressions to create the probability distributions
+            # for a single regression beta[i] is the tau-th (e.g. median or 30th percentile)
+            # for where dummy variable position i is equal to 1
+            # since we are fitting many quantile regressions at the same time, our beta is
+            # beta[tau, i] where tau stretches from 0.01 to 0.99
+            qr_lower = QuantileRegressionSolver()
+            qr_lower.fit(x_train_aug, delta_aug_lb, self.taus_lower, fit_intercept=False)
+            betas_lower = qr_lower.coefficients
+
+            qr_upper = QuantileRegressionSolver()
+            qr_upper.fit(x_train_aug, delta_aug_ub, self.taus_upper, fit_intercept=False)
+            betas_upper = qr_upper.coefficients
+
+            betas = np.concatenate([betas_lower, betas_upper])
+
+            # for each strata, we take the betas that belong to that stratum
+            # ie. for stratum [0, 1, 0] we take the betas (there is one for each tau between 0.01, 0.99])
+            # at position 1 (0-indexed)
+
+            # get all the betas for where x_stratum has a 1 (ie [1, 0, 0] position 0, [0, 1, 0] position 1 etc.)
+            betas_stratum = betas[:, np.where(x_stratum == 1)[0]].sum(axis=1)
+
+            # for this stratum value create ppf
+            stratum_ppfs_delta[tuple(x_stratum)] = ppf_creator(betas_stratum, self.taus, lb, ub)
+
+            # for this stratum value create cdf
+            stratum_cdfs_delta[tuple(x_stratum)] = cdf_creator(betas_stratum, self.taus)
+
+        return stratum_ppfs_delta, stratum_cdfs_delta
```

Comment:
We can't really guarantee that but we would have to be incredibly unlucky for this not to happen / this would be an indication that something profound on the model has failed and we probably shouldn't be using this model in the first place. In some sense though the guarantee comes from here: https://github.com/washingtonpost/elex-live-model/blob/conformal-bootstrap-model/src/elexmodel/models/BootstrapElectionModel.py#L1108 the lower/upper intervals are subtracted/added to the point prediction.